visa-cal: login and fetch based on dates

esakal opened the pull request:
# Motivations
Visa cal changed the API we were using in a way that introduced too many issues.

They removed essential information making it impossible to support the API we are using.

In this PR, I'm creating a new scraper using the puppeteer library. I track the status in the list below:
- [x] login into the account
- [x] filter transactions based on the start date
- [ ] support multiple accounts
- [ ] scraper transactions and convert them to the structure used by the library API
- [ ] fix lint issues and finetune the code

closes #603, closes #602

esakal commented:
@tpseven-777 @brafdlog @erikash @ifeins I completed the visa-cal scraper.

Feel free to take it to a spin and update how it went. Please read before the PR description above.

Thanks
Eran

@eshaham  fyi

erikash commented:
@esakal I think that option 2) is cleaner as I think the semantics of "one transaction per installment" are simpler and less error prone but since this requires changing the selectors I didn't go ahead with the fix yet..

So LMK what you think :)

esakal commented:
@erikash  and All,
Note that I pushed a fix to the amount and basically duplicate it by -1. I figure it out while importing to YNAB and getting a lot of refunds instead of payments :)

erikash commented:
Hi,
I’ve already fixed it in my PR, hopefully Eran will merge it soon :)

On Sun, 10 Oct 2021 at 20:14 sushi30 ***@***.***> wrote:

> Hi. I am using this branch and it appears it is not respecting the
> startDate parameter. Trying to find the cause but no success.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eshaham/israeli-bank-scrapers/pull/609#issuecomment-939518669>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABCIPL4JALCPZVSK27PYT3UGHCYPANCNFSM5EPQN57A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

esakal commented:
@sushi30 - @erikash  and I are almost there; we still work on the scraper; it is not an intuitive one.  I believe we are there; you can try it now or wait for us to finalize it in the coming days.

@erikash, I merged your changes into the pull request and fixed some issues:
1. the processedDate was incorrect; I changed it as we discussed earlier today.
2. we were missing a step of filtering transactions based on the start date; I took it from another place. Note that it considers the combineInstallments option, and I cannot really test it as I don't have installments, so test it carefully.
3. I added handling for the secondary table where the processed date is specific per row.
4. I mentioned us in the readme file.

Please pull the PR again, run it, and let us discuss more tomorrow.

sushi30 commented:
You can merge. I will create a separate PR.

On Thu, 14 Oct 2021 at 10:12, Eran Sakal ***@***.***> wrote:

> @sushi30 <https://github.com/sushi30> awesome, keep us in the loop about
> it.
> Do you want that we will wait until you will complete the multiple
> accounts scrapping before merging or you will do it separately?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eshaham/israeli-bank-scrapers/pull/609#issuecomment-943167321>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGJAQRPXSSKXHGWTUL4SLFLUG2NGRANCNFSM5EPQN57A>
> .
>

esakal commented:
@eshaham what is needed to merge it today or tomorrow?

As I wrote in previous messages this one introduced many design challenges and we needed to do some compromises but it seems that @erikash  @sushi30 and I are satisfy with the new scraper and the choices we did.

Next tasks:
- [ ] @eshaham please prioritize it and if possible approve PR
- [ ] @sushi30 once merge to master,  create a followup PR for the multiple cards support
- [ ] @erikash - let's also sync after the merge to tackle the change password scenario in its own PR

esakal commented:
@sushi30 @erikash @ifeins @eshaham  I changed the PR to draft mode for now. it seems that the new scraper stopped working again. I think that the VisaCal server added logic to detect the headless mode and block the request, I'm not sure why it is that important for them to prevent us from automating the process on a data that is ours to begin with.

![CleanShot 2021-10-17 at 16 20 29](https://user-images.githubusercontent.com/4751797/137628907-d64b8898-a01e-4bd2-bf66-810a693c0233.jpg)

Alternatives:
1. to try and figure out what is the information VisaCal wants us to send so they will be happy and let us automate the process
2. to contact them and see if this is that important for them to block the automation.
3. to add a disclaimer in our library that scrapping VisaCal is available only when not using an headless browser.

I might be mistaken but this is what I came up with till now.  

To reproduce it on your machines try to scrape with and without `showBrowser`

eshaham commented on src/scrapers/visa-cal.ts:
```diff
@@ -1,354 +1,302 @@
-        errorType: ScraperErrorTypes.AccountBlocked,
-      };
-    }
+// async function redirectOrDialog(page: Page): Promise<any> {
+//   return Promise.race([
+//     waitForNavigation(page),
+//     (async () => {
+//       try {
+//         await waitUntil(async () => {
+//           return hasInvalidPasswordError(page);
+//         }, 'wait for concrete error message', 10000, 1000);
+//       } catch (e) {
+//         // this is a valid scenario, waitUntil will fail once promise.race will handle the first promise
+//       }
+//     })(),
+//   ]);
+// }
```
@esakal can we delete this?

eshaham commented on src/scrapers/visa-cal.ts:
```diff
@@ -1,354 +1,302 @@
-    }
+  const accountNumber = /\d+$/.exec(accountId.trim())?.[0] ?? '';
+  accounts.push(await fetchTransactionsForAccount(page, startDate, accountNumber, scraperOptions));
+  // }
```
@esakal can we remove this?

sushi30 commented:
created a PR that works for me
